File: ReLeetCode/0543.py
class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        self.ans = 0
        self.diameterOfBinaryTreeHelper(root, self.ans)
        return self.ans

    def diameterOfBinaryTreeHelper(self, root: TreeNode, ans: list) -> int:
        if root == None:
            return 0
        left_max = self.diameterOfBinaryTreeHelper(root.left, self.ans)
        right_max = self.diameterOfBinaryTreeHelper(root.right, self.ans)
        self.ans = max(left_max + right_max, self.ans)
        return 1 + max(left_max, right_max)


# 明确最大直径的定义, 对于每个节点来说就是它的左子树高度 + 右子树高度.
# 所以可以在求树高度的递归过程中,记录每个节点的最大直径, 最后返回即可.

# 时间复杂度 O(n)
# 空间复杂度 O(lgn)

File: CppLeetCode/0543.cpp
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        helper(root);
        return max_res;
    }

    int helper(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        int left_height = helper(root->left);
        int right_height = helper(root->right);
        if (left_height + right_height > max_res) {
            max_res = left_height +right_height;
        }
        return 1 + max(left_height, right_height);
    }
private:
    int max_res = INT_MIN;
};

// 时间复杂度 O(n)
// 空间复杂度 O(h) h 树的高度

// 先求左子树的树高，然后再求右子树的树高，直径为 left_height + right_height
// 同时返回以当前节点为根的树的树高，也就是 1 + max(left_height, right_height)
// 这个算法遍历了树中的每一个节点，保证了求的不仅仅是经过 root 节点的直径。